### ISSUE TITLE
Could not find compiler set in environment variable CXX 

### ISSUE THREAD

**Issue by bbarnes52** [bazelbuild/rules_foreign_cc#13]
Setting and exporting the env variable CXX does not resolve the issue.

System info:
MacBook Pro (15-inch, 2017)
OS: High Sierra 10.13.1
Processor: 3.1 GHz Intel Core i7
Memory: 16 GB 2133 MHz LPDDR3

Error message:
ERROR: /Users/brian/cloud/roboflow/omg/BUILD:20:1: error executing shell command: 'set -e
export EXT_BUILD_ROOT=$(pwd)
source external/rules_foreign_cc/tools/build_defs/utils.sh
echo "Building external library 'pybind11'"
export TMPDIR=$(mktemp -d)
trap "{ rm -rf $TMPDIR; }" EXIT...' failed (Exit 1)
Building external library 'pybind11'
INSTALLDIR: "/private/var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/sandbox/darwin-sandbox/1/execroot/__main__/bazel-out/darwin-fastbuild/genfiles/roboflow/omg/pybind11"
EXT_BUILD_DEPS: ""
EXT_BUILD_ROOT: "/private/var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/sandbox/darwin-sandbox/1/execroot/__main__"
PATH: "/bin:/Users/brian/git/bazel/bazel-bin/src:/Users/brian/google-cloud-sdk/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/bin"
/var/folders/sg/_2yjj6v54lbckvqwr3jc4zdc0007t6/T/tmp.Vn8w1XYI /private/var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/sandbox/darwin-sandbox/1/execroot/__main__
-- Configuring incomplete, errors occurred!
See also "/var/folders/sg/_2yjj6v54lbckvqwr3jc4zdc0007t6/T/tmp.Vn8w1XYI/CMakeFiles/CMakeOutput.log".
mktemp: illegal option -- -
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
       mktemp [-d] [-q] [-u] -t prefix 
CMake Error at /usr/local/Cellar/cmake/3.12.0/share/cmake/Modules/CMakeDetermineCXXCompiler.cmake:47 (message):
  Could not find compiler set in environment variable CXX:

  
  /private/var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/sandbox/darwin-sandbox/1/execroot/__main__/external/local_config_cc/cc_wrapper.sh.

Call Stack (most recent call first):
  CMakeLists.txt:19 (project)

CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
Target //roboflow/omg:omg failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.638s, Critical Path: 0.17s
INFO: 0 processes.

**Comment by bbarnes52** [comment]
/private/var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/sandbox/darwin-sandbox/1/execroot/main/external/local_config_cc/cc_wrapper.sh does not exist.

it looks like the actual path should be /private/var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/execroot/main/external/local_config_cc/cc_wrapper.sh

**Comment by irengrig** [comment]
Thank you for the feedback!
Sorry for the delay, took me some time to figure it out...

The thing is, the toolchain files should be available to cmake_external, when it is executed (in the common case, in sandbox). 
So you should additionally use the option "additional_tools", which accepts list of labels of any tools that will be used by the build.

Here is the example for passing the Android NDK sources, so that cmake_external could use the compiler and linker from Android toolchain: [https://github.com/bazelbuild/rules_foreign_cc/blob/master/examples/cmake_android/BUILD](https://github.com/bazelbuild/rules_foreign_cc/blob/master/examples/cmake_android/BUILD)

Please let me know if that helps.

I am planning to indicate this problem better in documentation, and probably in error handling, keeping open for now.

**Comment by irengrig** [comment]
You should put a target into a list-value of additional_tools attribute, which contains this cc_wrapper.sh file (and all other files that it refers to).
It there is no such target yet, can you create a filegroup with the needed files?

**Comment by bbarnes52** [comment]
Would love to help, but frankly do not possess the know how to implement that.

**Comment by bbarnes52** [comment]
SGTM, thank you!

**Comment by bbarnes52** [comment]
Different error now ... looks like it might just be an issue with syntax? 

ERROR: /Users/brian/cloud/roboflow/omg/BUILD:20:1: error executing shell command: 'set -e
export EXT_BUILD_ROOT=$(pwd)
source external/rules_foreign_cc/tools/build_defs/utils.sh
echo "Building external library 'pybind11'"
export TMPDIR=$(mktemp -d)
trap "{ rm -rf $TMPDIR; }" EXIT...' failed (Exit 1)
Building external library 'pybind11'
INSTALLDIR: "/private/var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/sandbox/darwin-sandbox/1/execroot/__main__/bazel-out/darwin-fastbuild/genfiles/roboflow/omg/pybind11"
EXT_BUILD_DEPS: ""
EXT_BUILD_ROOT: "/private/var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/sandbox/darwin-sandbox/1/execroot/__main__"
PATH: "/bin:/Users/brian/git/bazel/bazel-bin/src:/Users/brian/git/bazel/bazel-bin/src/bazel:/Users/brian/google-cloud-sdk/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"
/var/folders/sg/_2yjj6v54lbckvqwr3jc4zdc0007t6/T/tmp.ypzAFRiQ /private/var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/sandbox/darwin-sandbox/1/execroot/__main__
-- The CXX compiler identification is AppleClang 9.1.0.9020039
-- Check for working CXX compiler: /var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/sandbox/darwin-sandbox/1/execroot/__main__/external/local_config_cc/cc_wrapper.sh
-- Check for working CXX compiler: /var/tmp/_bazel_brian/773e68ba09138a3e0d933a215b2b78a9/sandbox/darwin-sandbox/1/execroot/__main__/external/local_config_cc/cc_wrapper.sh -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /usr/bin/python (found version "2.7.10") 
-- Found PythonLibs: /System/Library/Frameworks/Python.framework/Versions/2.7/lib/libpython2.7.dylib
-- Performing Test HAS_CPP14_FLAG
-- Performing Test HAS_CPP14_FLAG - Success
-- Setting tests build type to MinSizeRel as none was specified
-- Building tests WITHOUT Eigen
-- Could NOT find Boost
-- Performing Test HAS_FLTO_THIN
-- Performing Test HAS_FLTO_THIN - Success
-- LTO enabled
-- Configuring incomplete, errors occurred!
See also "/var/folders/sg/_2yjj6v54lbckvqwr3jc4zdc0007t6/T/tmp.ypzAFRiQ/CMakeFiles/CMakeOutput.log".
mktemp: illegal option -- -
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
       mktemp [-d] [-q] [-u] -t prefix 
CMake Error at tests/CMakeLists.txt:194 (message):
  Running the tests requires pytest.  Please install it manually (try:
  /usr/bin/python -m pip install pytest)


Target //roboflow/omg:omg failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 2.254s, Critical Path: 2.15s
INFO: 0 processes.
FAILED: Build did NOT complete successfully

**Comment by irengrig** [comment]
Wow, that's interesting example!
Thank you for giving it another chance :)

As I see in the CMakeLists.txt of the pybind11 library, it requires pytest library (and that means that we should pass it somehow to the build rule).

However, there is an option to skip testing in build, pass the following attribute in cmake_external:
cmake_options = ["-DPYBIND11_TEST=False"],

Unfortunately, this is not the end of the story. 
pybind11 itself defines some CMake modules, which should be used by client CMakeLists.txt files to build pythin-invocable C++ code.

I made modifications for passing CMake modules up the chain of libraries, but I want to verify the solution better before submitting here.
Currently you can have a look in this pull request: #34.

**Comment by irengrig** [comment]
I added an example of building .cpp file with pybin11: #43.
Unfortunately, I did not figure it out yet how to correctly call it from Python :)
At least I got the result .so file.
Hope this helps.
This pull request: #34 is not relevant any more, closing it.

**Comment by irengrig** [comment]
Yep, of course you can build with pybind11 without CMake, there are examples how to do that on their site.
